Utils/src/HiddenSectorProducer.cc

```diff
@@ -22,143 +27,436 @@
+        }
+        else if (isParticle(DarkQuarkIDs_,dau)) firstQd.insert(dau);
+        else if (isParticle(DarkGluonIDs_,dau)) firstGd.insert(dau);
+      }
+    }
+  }
+}
+
+int HiddenSectorProducer::checkLast(const reco::GenJet& jet, const CandSet& stableDs, int value, double& frac) const {
+  //compare jet constituents to set of last particles
+  //also compute dark pt fraction in same loop
+  bool match = false;
+  LorentzVector p4;
+  LorentzVector totPt = jet.p4();
+  for(unsigned i = 0; i < jet.numberOfDaughters(); ++i){
+    CandPtr dau = jet.daughter(i);
+    if(isAncestor(DarkHadronIDs_,dau)){
+        match = true;
+        p4 += dau->p4();
+    }
+  }
+  for(const auto& part : stableDs){
+     if(reco::deltaR(jet, *part) < coneSize_)
+     {
+       p4 += part->p4();
+       totPt += part->p4();
+     }
+  }
+  frac = p4.pt()/totPt.pt();
+  return match ? value : 0;
+}
+
+int HiddenSectorProducer::checkFirst(const reco::GenJet& jet, const CandSet& firstP, int value) const {
+  //compare first particles using jet cone (can't compare constituents' parents because all dark hadrons descend from all first dark particles, strong force problems)
+  for(const auto& part : firstP){
+    if(reco::deltaR(jet, *part) < coneSize_) return value;
+  }
+  return 0;
+}
+
+// match particles from mediator to jets
+void HiddenSectorProducer::matchPFMtoJet(CandPtr part, std::vector<const reco::GenJet*>& matchedJets, const reco::GenJet& jet, int& nPartPerJet, int& t_MT2JetID, int mt2ID) const {
+  if(reco::deltaR(jet,*part) < coneSize_)
+  {
+    matchedJets.push_back(&jet);
+    nPartPerJet ++;
+    t_MT2JetID = mt2ID;
+  }
+}
+
+double HiddenSectorProducer::calculateMT2(edm::Handle<edm::View<reco::GenMET>> h_genmets, const reco::GenJet*& dQM1J, const reco::GenJet*& SMM1J, const reco::GenJet*& dQM2J, const reco::GenJet*& SMM2J) const {
+  const auto& i_met = h_genmets->front();
+  double METx = i_met.px();
+  double METy = i_met.py();
+  LorentzVector FJet0 = dQM1J->p4() + SMM1J->p4();
+  LorentzVector FJet1 = dQM2J->p4() + SMM2J->p4();
+  return asymm_mt2_lester_bisect::get_mT2(
+    FJet0.M(), FJet0.Px(), FJet0.Py(),
+    FJet1.M(), FJet1.Px(), FJet1.Py(),
+    METx, METy, 0.0, 0.0, 0
+  );
+}
+
+double HiddenSectorProducer::calculateMT2(edm::Handle<edm::View<reco::GenMET>> h_genmets, const LorentzVector& dQM1J, const LorentzVector& SMM1J, const LorentzVector& dQM2J, const LorentzVector& SMM2J) const {
+  const auto& i_met = h_genmets->front();
+  double METx = i_met.px();
+  double METy = i_met.py();
+  LorentzVector FJet0 = dQM1J + SMM1J;
+  LorentzVector FJet1 = dQM2J + SMM2J;
+  return asymm_mt2_lester_bisect::get_mT2(
+    FJet0.M(), FJet0.Px(), FJet0.Py(),
+    FJet1.M(), FJet1.Px(), FJet1.Py(),
+    METx, METy, 0.0, 0.0, 0
+  );
+}
+
+const reco::GenJet* HiddenSectorProducer::highPt(std::vector<const reco::GenJet*>& Js) const {
+  double largestPt = 0;
+  const reco::GenJet* largestPtJ;
```

Comment:
initialize this to `nullptr` (there was a compiler warning: `warning: 'largestPtJ' may be used uninitialized in this function [-Wmaybe-uninitialized]`)